### DISCUSSION TITLE
want to access the properties in the custom action during uninstallation

### DISCUSSION THREAD

**Discussion by surajk46** [oleg-shilo/wixsharp#1689] [Q&A]
Hi Oleg,
 
I want to access the properties in the custom action during uninstallation. In my project with managedUI of windows form, I have a Windows Forms TreeView to select the features, and I populate the properties to determine whether a feature is selected or not.
 
While installing, I can retrieve the selected properties and, based on this, I have added conditions in the custom action to install , and it executes accordingly and works as expected.
 
However, while uninstalling the product, I can get the selected feature value in the BeforeInstall event like this:
 
```
if (e.IsUninstalling)
{
	if (e.Session.Features.Contains("FEATURE1") && e.Session.Features["FEATURE1"].CurrentState == InstallState.Local) 
	{
   		e.Session["IS_FEATURE1_SELECTED"] = "TRUE"; 
	}
	else
	{
		e.Session["IS_FEATURE1_SELECTED"] = "FALSE"; 
	}
}
```

 
This works as expected, but when I try to use the value in the uninstallation custom action like this:
 
```
var UnInstall_FEATURE1 = new ElevatedManagedAction(CustomActions.UnInstall_FEATURE1, Return.check, When.Before, Step.RemoveFiles, Condition.BeingUninstalled);
UnInstall_FEATURE1.UsesProperties = $"ADDFEATURES=[ADDFEATURES], ADDLOCAL=[ADDLOCAL], {IS_FEATURE1_SELECTED}=[{IS_FEATURE1_SELECTED}]";
```
 
I am getting a blank values for all property in the custom action during **uninstallation**. Even after following this approach, the value is not being passed correctly.

**Comment by oleg-shilo** [comment]
Correct. It is an unfortunate limitation of MSI architecture. At that time the MSI session object is closed and none of the properties can be accessed.

The approach you taken is correct but I am not sure why are you substituting the property name with it's value. Unless you have your specific reason....
I expected the code to be something like this:

```C#
UnInstall_FEATURE1.UsesProperties = $"ADDFEATURES=[ADDFEATURES], ADDLOCAL=[ADDLOCAL], IS_FEATURE1_SELECTED=[IS_FEATURE1_SELECTED]";
``` 

Thou if I were facing the same task I would go with a simpler syntax:
```C#
project.DefaultDeferredProperties += ",IS_FEATURE1_SELECTED,ADDFEATURES,ADDLOCAL";
```
And utilize the events. Way more natural to work with. But this is really a matter of preferences :)
